EventFilter/Phase2TrackerRawToDigi/plugins/DTHDAQToFEDRawDataConverter.cc

```diff
@@ -0,0 +1,145 @@
+#include "FWCore/Framework/interface/stream/EDProducer.h"
+#include "FWCore/Framework/interface/Event.h"
+#include "FWCore/Framework/interface/MakerMacros.h"
+#include "DataFormats/FEDRawData/interface/FEDRawData.h"
+#include "DataFormats/FEDRawData/interface/FEDRawDataCollection.h"
+#include "FWCore/Utilities/interface/Exception.h"
+#include <fstream>
+#include <vector>
+#include <iostream>
+#include <iomanip>
+#include <cstdint>
+
+// Constants for bit field widths, markers, and sizes
+constexpr unsigned int orbitHeaderSize = 32;
+constexpr unsigned int fragmentTrailerSize = 16;
+constexpr unsigned int fragmentPayloadWordSize = 16;  // Each fragment payload word is 16 bytes
+constexpr uint8_t orbitHeaderMarkerH = 0x48;
+constexpr uint8_t orbitHeaderMarkerO = 0x4F;
+constexpr uint8_t fragmentTrailerMarkerF = 0x48;
+constexpr uint8_t fragmentTrailerMarkerH = 0x46;
+
+class DTHDAQToFEDRawDataConverter : public edm::stream::EDProducer<> {
+public:
+  explicit DTHDAQToFEDRawDataConverter(const edm::ParameterSet&);
+  ~DTHDAQToFEDRawDataConverter() override = default;
+
+  void produce(edm::Event&, const edm::EventSetup&) override;
+
+private:
+  std::string inputFile_;
+  unsigned int fedId_;
+
+  std::vector<char> readRawFile(const std::string& inputFile);
+  void parseAndDumpEventData(const std::vector<char>& buffer);
+  void reverseParseFragments(const std::vector<char>& buffer, size_t startIdx, uint32_t packetWordCount, uint16_t eventCount);
+  void printHex(const std::vector<char>& buffer, size_t length);
+};
+
+DTHDAQToFEDRawDataConverter::DTHDAQToFEDRawDataConverter(const edm::ParameterSet& config)
+  : inputFile_(config.getParameter<std::string>("inputFile")),
+    fedId_(config.getParameter<unsigned int>("fedId")) {
+  produces<FEDRawDataCollection>();
+}
+
+std::vector<char> DTHDAQToFEDRawDataConverter::readRawFile(const std::string& inputFile) {
```

Comment:
This function is returning a very large object, without using references, so this requires a big copy, which is not efficient. I don't know if there's a more efficient way to write this?

Reply:
I tried using `return std::move(buffer);  // Ensure move semantics` but the complier did not like that
![Screenshot 2024-12-09 at 10 24 56 AM](https://github.com/user-attachments/assets/a3265409-024b-4503-a944-7d94ad62eb11)